packages/xrpc-server/src/rate-limiter.ts

```diff
@@ -142,48 +129,128 @@ export const formatLimiterStatus = (
   }
 }
 
-export const consumeMany = async (
-  ctx: XRPCReqContext,
-  fns: RateLimiterConsume[],
-): Promise<RateLimiterStatus | RateLimitExceededError | null> => {
-  if (fns.length === 0) return null
-  const results = await Promise.all(fns.map((fn) => fn(ctx)))
-  const tightestLimit = getTightestLimit(results)
-  if (tightestLimit === null) {
-    return null
-  } else if (tightestLimit instanceof RateLimitExceededError) {
-    setResHeaders(ctx, tightestLimit.status)
-    return tightestLimit
-  } else {
-    setResHeaders(ctx, tightestLimit)
-    return tightestLimit
+/**
+ * Wraps a {@link RateLimiterI} instance with custom key and points calculation
+ * functions.
+ *
+ * @note noop if both calcKey and calcPoints are undefined.
+ */
+export function updateRateLimiterOptions(
+  rateLimiter: RateLimiterI,
+  {
+    calcKey,
+    calcPoints,
+  }: {
+    calcKey?: CalcKeyFn
+    calcPoints?: CalcPointsFn
+  } = {},
+): RateLimiterI {
+  if (!calcKey && !calcPoints) return rateLimiter
+
+  return {
+    consume: async (ctx, opts) =>
+      rateLimiter.consume(ctx, {
+        calcKey: opts?.calcKey ?? calcKey,
+        calcPoints: opts?.calcPoints ?? calcPoints,
+      }),
+    reset: async (ctx, opts) =>
+      rateLimiter.reset(ctx, {
+        calcKey: opts?.calcKey ?? calcKey,
+      }),
   }
 }
 
-export const resetMany = async (
-  ctx: XRPCReqContext,
-  fns: RateLimiterReset[],
-): Promise<void> => {
-  if (fns.length === 0) return
-  await Promise.all(fns.map((fn) => fn(ctx)))
+/**
+ * Combines multiple rate limiters into one.
+ *
+ * The combined rate limiter will return the tightest (most restrictive) of all
+ * the provided rate limiters.
+ */
+function combine(
+  input: Iterable<RateLimiterI | null | undefined>,
+): RateLimiterI | undefined {
+  const rateLimiters = Array.from(input).filter((rl) => rl != null)
+
+  if (rateLimiters.length === 0) return undefined
+  if (rateLimiters.length === 1) return rateLimiters[0]
+
+  return {
+    consume: async (ctx, opts) => {
+      const promises: ReturnType<RateLimiterConsume>[] = []
+      for (const rl of rateLimiters) promises.push(rl.consume(ctx, opts))
+      return Promise.all(promises).then(getTightestLimit)
+    },
+    reset: async (ctx, opts) => {
+      const promises: ReturnType<RateLimiterReset>[] = []
+      for (const rl of rateLimiters) promises.push(rl.reset(ctx, opts))
+      await Promise.all(promises)
+    },
+  }
+}
+
+export type RouteRateLimiterOptions = {
+  bypass?: (ctx: XRPCReqContext) => boolean
```

Comment:
Hidden motive: remove dependency on Express by moving Express dependent logic (`req.ip`) outside of this module.